### DISCUSSION TITLE
Stripping down browser forwarding with external Chrome stack

### DISCUSSION THREAD

**Discussion by klzgrad** [v2fly/v2ray-core#2092] [Ideas]
From what I can tell browser forwarding is resource intensive and requires user maintenance of the forwarding page. This makes it difficult to deploy on embedded devices.

Here is an idea: Chrome's network stack can be stripped down and built into an executable that accepts an inbound socks5 or other type of connection and translates that into an outbound TLS connection or websocket connection using Chrome's network stack.

A previous similar attempt was with SagerNet which tried to package Chrome net stack via the Cronet interface into Cronet-Go, so that it is possible to link Cronet into a single Go executable. That attempt was not very successful because it requires CGO and CGO is problematic for portable builds.

The proposal this time is to have the Chrome net stack running as an external process and communicate with it via sockets or some kind of IPC. The Chrome net stack executable will be built separately without interference with Go build process, will cover most architectures that Go supports, and will be fairly small, ~5MB uncompressed. This would be superior to full browser forwarding as it is much smaller and can be packaged and deployed automatically.

Most of what's described here already exists as elements in naiveproxy so it takes no much effort to create this executable. Changes needed on the v2fly side are related to protocol adaptation and some integration work at the existing browser forwarding level.

What's your opinion?

**Comment by AkinoKaede** [comment]
Good idea. Can we make the executable be made to receive a protocol with destination, protocol type (TLS/WebSocket) and stream (e.g. VMess)?

**Reply by klzgrad** [reply]
I'll see what I can make out of this.

**Reply by AkinoKaede** [reply]
> I'll see what I can make out of this.

Thanks.

**Reply by klzgrad** [reply]
As for the protocol, my design so far is to build on the http connect method to create a bidirectional stream socket, say, chrome-net.exe listening on localhost:8080 accepting a request like this

```
CONNECT example.com:443 HTTP/1.1
Tunnel: tls
Tunnel-Params: ...

HTTP/1.1 200 OK
```

After this preamble, chrome-net.exe creates a TLS socket to example.com:443 and forward tcp byte stream within localhost after the handshake to example.com over the TLS layer. Consider this as `openssl s_client` with parameters specified in the preamble. This could combine with Go http or ws to create https/1.1 and wss, but probably is not recommended for h2c because Go h2 and chrome h2 have some discrepancies.

The `Tunnel` header can also be wss, which creates a wss tunnel to presumably Nginx and its backend.

It can probably be h2 to create some kind of h2 tunnel (multiplexed), but I'm not sure if it is to use the default http connect tunnel or to support user defined headers (which requires some header translation from h1 to h2)? But this is also expected to work with a h2 reverse proxy that demuxes the h2 tunnel to h1 backend.

The `Tunnel-Params` header, or it can be split into several headers, controls the tunnel parameters. I can think of several things but there could be more: alpn, fast open, early data, network isolation key.

> stream (e.g. VMess)

I do not understand what is this. There is surely no C++ implementation of VMess in Chromium and it would not be possible to offload VMess processing to Chrome stack, unlike tls and ws.

**Reply by AkinoKaede** [reply]
That's a good idea. Stram means TCP steam, and you said it is supported.
